source/mqtt/Mqtt5Client.cpp

```diff
@@ -389,20 +414,59 @@ namespace Aws
                 }
 
                 toSeat = new (toSeat) Mqtt5Client(options, allocator);
-                return std::shared_ptr<Mqtt5Client>(
+                // Creation failed, make sure we release the allocated memory
+                if (!*toSeat)
+                {
+                    Crt::Delete(toSeat, allocator);
+                    return nullptr;
+                }
+
+                std::shared_ptr<Mqtt5Client> shared_client = std::shared_ptr<Mqtt5Client>(
                     toSeat, [allocator](Mqtt5Client *client) { Crt::Delete(client, allocator); });
+                shared_client->m_selfReference = shared_client;
+                return shared_client;
             }
 
             Mqtt5Client::operator bool() const noexcept { return m_client != nullptr; }
 
             int Mqtt5Client::LastError() const noexcept { return aws_last_error(); }
 
-            bool Mqtt5Client::Start() const noexcept { return aws_mqtt5_client_start(m_client) == AWS_OP_SUCCESS; }
+            bool Mqtt5Client::Start() noexcept
+            {
+                if (aws_rw_lock_try_rlock(&m_client_lock) != AWS_ERROR_SUCCESS || m_client == nullptr)
```

Comment:
This pattern doesn't unlock a taken read lock if the client is null.

Maybe we should create a simple scoped rw lock class where creation fatal-assert-takes the lock and destruction releases it.

Also you can't release the lock until the C function returns.  Close() could have decrefed it inbetween unlock and aws_mqtt5_* invocation.

These apply to all usages of this pattern.